### PR TITLE
RRF: Support M291 dialogs and interactive macros

### PR DESCRIPTION
**Pull request by pfn**
### Requirements

No new requirements

### Description

Enables support for M291 dialogs when initiated outside of the terminal menu (DWC, macros, printing gcode, etc.). S0, S1, S2 and S3 are supported modes (although for S0, the OK button is always shown). X/Y/Z arguments are not supported.

* Add RRFStatusControl.c, tracks state of RRF and launches M408 requests
* Refactor `fanQuery` into `rrfStatusQuery()`
* Update `request_M98` macro execution to use RRFStatusControl
* Block sending `M27` and `M408` while inside of the terminal menu
* Process `M291` information from `M408` responses
* Add status call `rrfStatusIsMacroBusy`
* Prevent `M27`, `M105` and `M155` from running while macro busy
* Skip waiting for `ok` response for `M98`
* Skip waiting for `ok` response on `M292` and `M408` while macro busy
* Remove prompt to start a macro. Macros will run on click.
  * dangerous macros should have their own M291 prompt
* Updates the running macro display

### Benefits

* M291 dialog support
* Interactive macros

### Testing

This has been tested and verified to work on my MKS TFT28

### PR Status

This is ready to merge pending any reviews.

### CODE REVIEW

**Comment by pfn** [comment]
Oh, duh. M98 is a blocking gcode.

**Comment by digant73** [comment]
it's better you rename parseAckJson to RRFParseAckJson as done for all the other RRF specific files.
Also move RRFMacrosParser to folder API instead of Menu. It's not a menu

**Comment by pfn** [comment]
> it's better you rename parseAckJson to RRFParseAckJson as done for all the other RRF specific files.
> Also move RRFMacrosParser to folder API instead of Menu. It's not a menu

done @digant73

File: TFT/src/User/API/RRFStatusControl.c [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+void rrfStatusSet(char status)
+{
+  rrf_status = status;
+  if (status != 'B')
+  {
+    macro_busy = false;
+  }
+}
+
+void rrfStatusSetBusy(void)
```
Comment (digant73) [inline TFT/src/User/API/RRFStatusControl.c:24]: make all of those functions made by only one line as staic inline. E.g. "static inline void rrFStatusSet etc...) it is more efficient and uses les flash space (in that case)

File: TFT/src/User/API/parseACK.c [hunk 2]
```diff
@@ -1,6 +1,6 @@
 #include "parseACK.h"
 #include "includes.h"
-#include "parseACKJson.hpp"
+#include "RRFParseACK.hpp"
```
Comment (digant73) [inline TFT/src/User/API/parseACK.c:3]: move this include in inludes.h

**Review by digant73 — CHANGES_REQUESTED** [review]

**Comment by digant73** [comment]
As you already are doing try to put all the logic related to RRF on functions (with rrf prefix). Provide those functions on specific RRF files (with RRF prefix). Always try to not change the existing code to adapt it to RRF but try to engage the rrf functions with all the specif logic. When the functions are called only once or the function are made be just one easy instruction, define them as "static inline" (it will uses less flash). When possible put RRF logic/functions as last choice (not first)

**Comment by digant73** [comment]
In RRFMacros.c I would also move #include "RRFMacrosParser.hpp" to includes.h

**Comment by bttguy** [comment]
This interactive macro support sounds great! Thanks a lot @pfn  ! :)

File: TFT/src/User/API/parseACK.c (hunk 2 above)
Comment (pfn) [reply TFT/src/User/API/parseACK.c:3]: I'm not a fan of this change, I feel that scope should be minimized, and only parseACK.c should have access to anything in RRFParseACK.

**Comment by pfn** [comment]
> In RRFMacros.c I would also move #include "RRFMacrosParser.hpp" to includes.h

@digant73 similar to what I said for RRFParseACK, I believe the same goes here. Only RRFMacros should have access to RRFMacrosParser.

File: TFT/src/User/API/parseACK.c (hunk 2 above)
Comment (digant73) [reply TFT/src/User/API/parseACK.c:3]: no problem if the scope will be limited only on parseAck

**Comment by digant73** [comment]
@pfn if possible have a look to #1856. It seams PID is not pefectly working yet

**Comment by pfn** [comment]
> @pfn if possible have a look to #1856. It seams PID is not pefectly working yet

@digant73 I'll give it a look if I have some time

**Comment by bttguy** [comment]
@pfn Is this PR adds support also for "menu/folder stucture" for macros? Right now if I categorize them on DWC, they only show up as a list on the TFT without any folders.

**Comment by pfn** [comment]
> @pfn Is this PR adds support also for "menu/folder stucture" for macros? Right now if I categorize them on DWC, they only show up as a list on the TFT without any folders.

@bttguy no, there is no change in that area. However, folders have been supported for quite some time now. I believe it was added by @Msq001 a while back.

**Comment by bttguy** [comment]
Hmm, I will check it and update the FR-s accordingly. I just checked on the RRF RF I created in #1791 if there is anything I can add/update. :)

**Pull request merged**